src/execution_plan/execution_plan_build/execution_plan_util.c

```diff
@@ -305,24 +331,80 @@ OpBase **ExecutionPlan_CollectOps
 	return ops;
 }
 
+// collect all operations belonging to the given plan
+// it is the callers responsibility to free the returned array
+OpBase **ExecutionPlan_CollectAllOps
+(
+	const ExecutionPlan *plan, // plan to collect ops from
+	uint *n                    // [output] number of ops collected
+) {
+	ASSERT(n    != NULL);
+	ASSERT(plan != NULL);
+
+	*n = 0;
+
+	// can't scan plan, return NULL
+	if(plan->root == NULL) {
+		return NULL;
+	}
+
+	OpBase **ops = rm_malloc(sizeof(OpBase*) * 1);
+
+	uint s = 0;  // current index into ops array
+	uint e = 1;  // length of ops array
+	ops[0] = plan->root;
+
+	// collect upwards using child pointers
+	while(s < e) {
+		OpBase *current = ops[s++];
+
+		// try adding current children
+		for(uint i = 0; i < OpBase_ChildCount(current); i++) {
+			OpBase *child = OpBase_GetChild(current, i);
+
+			// add child only if it belong to given plan
+			if(child->plan == plan) {
+				ops = rm_realloc(ops, sizeof(OpBase*) * (e + 1));
+				ops[e++] = child;
+			}
+		}
+	}
+
+	// collect downwards using parent pointer
+	OpBase *current = plan->root->parent;
+	while(current != NULL && current->plan == plan) {
+		ops = rm_realloc(ops, sizeof(OpBase*) * (e + 1));
+		ops[e++] = current;
+
+		// move on to the next parent
+		current = current->parent;
+	}
```

Comment:
it's unlikely that the plan's root has a parent which belongs to the same plan.
consider asserting!